### PR TITLE
Common > Fix JS error on dialog.

### PR DESCRIPTION
**Pull request by juanfra**
🎫 https://moderntribe.atlassian.net/browse/TCMN-34
🎥 https://www.loom.com/share/e65ccf5c62c745a8bc5e9ec41d538702

### CODE REVIEW

File: src/resources/js/dialog.js [hunk 1]
```diff
@@ -1,17 +1,17 @@
 var tribe = tribe || {};
 tribe.dialogs = tribe.dialogs || {};
-tribe.dialogs.dialogs = tribe.dialogs.dialogs || {};
+tribe.dialogs.dialogs = tribe.dialogs.dialogs || [];
 tribe.dialogs.events = tribe.dialogs.events || {};
```
Comment (sc0ttkclark) [inline src/resources/js/dialog.js]: Should this be an array too?

File: readme.txt [hunk 2]
```diff
@@ -2,6 +2,10 @@
 
 == Changelog ==
 
+= [4.11.2] TBD =
+
+* Fix - Fix JavaScript error in tribe dialog when there are no dialogs. Change fallback from object to array. [TCMN-34]
```
Comment (bordoni) [inline readme.txt]: Avoid duplication of the word "Fix" in the changelog.

File: src/resources/js/dialog.js [hunk 3]
```diff
@@ -1,17 +1,17 @@
...
 
-	( function ( obj ) {
+	( function( obj ) {
```
Comment (bordoni) [inline src/resources/js/dialog.js]: I know it's previous code but indentation is wrong.

File: src/resources/js/dialog.js [hunk 4]
```diff
@@ -1,17 +1,17 @@
 var tribe = tribe || {};
 tribe.dialogs = tribe.dialogs || {};
-tribe.dialogs.dialogs = tribe.dialogs.dialogs || {};
+tribe.dialogs.dialogs = tribe.dialogs.dialogs || [];
```
Comment (bordoni) [inline src/resources/js/dialog.js]: We should move the two definitions for `tribe.dialogs.dialogs` and `tribe.dialogs.events` to inside of the "module".

File: src/resources/js/dialog.js [hunk 5]
```diff
@@ -1,17 +1,17 @@
...
+	( function( obj ) {
 		'use strict';
 
 		document.addEventListener(
 			'DOMContentLoaded',
-			function () {
-				tribe.dialogs.dialogs.forEach(function(dialog) {
+			function() {
+				tribe.dialogs.dialogs.forEach( function( dialog ) {
 					var objName     = 'dialog_obj_' + dialog.id;
```
Comment (bordoni) [inline src/resources/js/dialog.js]: This should be a method that returns the `objName`.

File: src/resources/js/dialog.js [hunk 6]
```diff
@@ -1,17 +1,17 @@
...
-					window[objName] = new window.A11yDialog({
+					window[objName] = new window.A11yDialog( {
```
Comment (bordoni) [inline src/resources/js/dialog.js]: we should also store that variable inside of the `dialogs` object stored on the `obj`.
Also spacing around `[]`

File: src/resources/js/dialog.js [hunk 7]
```diff
@@ -1,17 +1,17 @@
...
 tribe.dialogs = tribe.dialogs || {};
-tribe.dialogs.dialogs = tribe.dialogs.dialogs || {};
+tribe.dialogs.dialogs = tribe.dialogs.dialogs || [];
 tribe.dialogs.events = tribe.dialogs.events || {};
 
-	( function ( obj ) {
+	( function( obj ) {
 		'use strict';
 
 		document.addEventListener(
```
Comment (bordoni) [inline src/resources/js/dialog.js]: if we are using jQuery here, we should use it here too.

File: src/resources/js/dialog.js [hunk 8]
```diff
@@ -1,17 +1,17 @@
...
 			'DOMContentLoaded',
-			function () {
```
Comment (bordoni) [inline src/resources/js/dialog.js]: This method should be named, something like `init`

File: src/resources/js/dialog.js [hunk 9]
```diff
@@ -24,26 +24,26 @@ tribe.dialogs.events = tribe.dialogs.events || {};
...
 						}
 
-						jQuery( tribe.dialogs.events ).trigger( dialog.closeEvent, [dialogEl, event] );
+						jQuery( tribe.dialogs.events ).trigger( dialog.closeEvent, [ dialogEl, event ] );
 					});
 				});
 			}
 		)
-	})(tribe.dialog);
+	})( tribe.dialog );
```
Comment (bordoni) [inline src/resources/js/dialog.js]: We should pass `jQuery` and use `$`

File: src/resources/js/dialog.js (hunk 1 above)
Comment (juanfra) [reply src/resources/js/dialog.js]: We don't use it as array, like we do with `tribe.dialogs.dialogs`: https://github.com/moderntribe/tribe-common/blob/97c251c6a8f1d72ab9545a339985de9032c57bbb/src/resources/js/dialog.js#L35

File: src/resources/js/dialog.js [hunk 10]
```diff
@@ -1,49 +1,71 @@
...
-			}
-		)
-	})(tribe.dialog);
+
+( function( $, obj ) {
+	'use strict';
+
+	var $document = $( document );
+	tribe.dialogs.dialogs = tribe.dialogs.dialogs || [];
+	tribe.dialogs.events = tribe.dialogs.events || {};
```
Comment (bordoni) [inline src/resources/js/dialog.js]: We should use `obj` since it's being passed down on the method.

File: src/resources/js/dialog.js [hunk 11]
```diff
@@ -1,49 +1,71 @@
...
+
+	/**
+	 * Initialize tribe dialogs.
+	 *
+	 * @since TBD
+	 *
+	 * @return {void}
+	 */
+	obj.init = function() {
+		tribe.dialogs.dialogs.forEach( function( dialog ) {
```
Comment (bordoni) [inline src/resources/js/dialog.js]: We should use `obj` since it's being passed down on the method.

File: src/resources/js/dialog.js [hunk 12]
```diff
@@ -1,49 +1,71 @@
...
+				wrapperClasses: dialog.wrapperClasses,
+			} );
+
+			window[ objName ].on( 'show', function( dialogEl, event ) {
+				if ( event ) {
+					event.preventDefault();
+					event.stopPropagation();
+				}
+
+				$( tribe.dialogs.events ).trigger( dialog.showEvent, [ dialogEl, event ] );
```
Comment (bordoni) [inline src/resources/js/dialog.js]: We should use `obj` since it's being passed down on the method.

File: src/resources/js/dialog.js [hunk 13]
```diff
@@ -1,49 +1,71 @@
...
+			} );
+
+			window[ objName ].on( 'hide', function ( dialogEl, event ) {
+				if ( event ) {
+					event.preventDefault();
+					event.stopPropagation();
+				}
+
+				$( tribe.dialogs.events ).trigger( dialog.closeEvent, [ dialogEl, event ] );
```
Comment (bordoni) [inline src/resources/js/dialog.js]: We should use `obj` since it's being passed down on the method.

File: src/resources/js/dialog.js [hunk 14]
```diff
@@ -1,49 +1,71 @@
...
+				contentClasses: dialog.contentClasses,
+				effect: dialog.effect,
+				effectEasing: dialog.effectEasing,
+				effectSpeed: dialog.effectSpeed,
+				overlayClasses: dialog.overlayClasses,
+				overlayClickCloses: dialog.overlayClickCloses,
+				trigger: dialog.trigger,
+				wrapperClasses: dialog.wrapperClasses,
+			} );
+
```
Comment (bordoni) [inline src/resources/js/dialog.js]: We should store a version of the variable inside of the dialog object. Like `dialog.a11yInstance` or something like that.

File: src/resources/js/dialog.js [hunk 15]
```diff
@@ -1,49 +1,71 @@
...
+	 * Get the dialog name.
+	 *
+	 * @since TBD
+	 *
+	 * @param {obj} dialog The dialog object
+	 *
+	 * @return {string} the dialog name.
+	 */
+	obj.getDialogName = function( dialog ) {
+		return 'dialog_obj_' + dialog.id;
```
Comment (bordoni) [inline src/resources/js/dialog.js]: Should we be prefixing this?

Comment (juanfra) [reply src/resources/js/dialog.js]: I feel we're moving far away from the original fix which was just replacing:

`tribe.dialogs.dialogs = tribe.dialogs.dialogs || {};`
with this:
`tribe.dialogs.dialogs = tribe.dialogs.dialogs || [];`

And we're reimplementing the dialogs, with the QA risk implications, given that I was not the one who implemented this. Maybe we can end up handling a re-implementation on a separate ticket or modifying the ticket so QA knows that they need to test not only the original thing but the entire dialogs.

Comment (sc0ttkclark) [reply src/resources/js/dialog.js]: Agree here

**Comment by juanfra** [comment]
@bordoni @sc0ttkclark I can't get this one merged, can any of you merge this one?

**Pull request merged**